### PR TITLE
 RHOAIENG-16076: tests(gha): run Makefile tests on opendatahub-io/notebooks Github Actions

### PR DESCRIPTION
**Pull request by jiridanek**
https://issues.redhat.com/browse/RHOAIENG-16076

This is building up on the preparatory work merged in
* https://github.com/opendatahub-io/notebooks/pull/780
* https://github.com/opendatahub-io/notebooks/pull/777
* https://github.com/opendatahub-io/notebooks/pull/782
* https://github.com/opendatahub-io/notebooks/pull/783

## Description

Deploys cri-o backed kubernetes with kubeadm to run our Makefile tests.

I considered using KinD, or Microshift, or Minikube, or OpenShift Local, or some other Kubernetes, but decided not to, because

* Microshift still requires Red Hat pull secret, so it's hard to run this for PRs
  * There is something new there, though https://github.com/openshift/microshift/tree/main/okd/src
* OpenShift Local takes long time to start, the OpenShift version needs pull-secret, and the OKD version is not getting much love
* KinD, etc. with CRI-O requires custom node image https://github.com/cri-o/cri-o/blob/release-1.31/tutorials/crio-in-kind.md#cri-o-in-kind

The reason why I need cri-o is that I want to share the podman-built images into the cluster without having to copy them. Copying images is difficult because the images can be huge.

* seems that k3d would've been also possible, https://devopstales.github.io/kubernetes/k3s-crio/

### Known issues that I do intend to fix

```
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... failed: Name or service not known.
wget: unable to resolve host address ‘raw.githubusercontent.com’
```

* [x] This needs a retry, because it happens from time to time. Also, we should not fetch test data from the main branch of the repo, we should use what's in the current checkout.

This went away after I redid my Kubernetes network. I had this problems with flannel, and since then I switched to a plain bridge with firewall masquerade. I think that (when I noticed this problem the first time, or maybe at some later point) my pods were completely unable to connect to the outside (also unable to connect to other pods, like the coredns pod, but I was dealing with other issues so I mistakenly thought the networking issues are a flakiness.

---

```
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f95ec277fd0>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/papermill/
ERROR: Could not find a version that satisfies the requirement papermill (from versions: none)
Notice:  A new release of pip is available: 24.2 -> 24.3.1
Notice:  To update, run: pip install --upgrade pip
ERROR: No matching distribution found for papermill
command terminated with exit code 1
```

* [x] Could it be that freshly started pod has spotty network and I need to wait?

Looks like the same problem as above. My pods were completely offline. The bridge solution needs no waiting!

### Known issues (that I don't intend to fix for this PR)

* GHA for a pull request skips intermediate images that have their own tests. Those tests will not be run, only leaf images in the dependency chain are having their tests run. (The push GHA tests everything.)

## How Has This Been Tested?

* https://github.com/jiridanek/notebooks/actions/runs/12064274658

## Merge criteria:

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: base/ubi9-python-3.11/Dockerfile
Comment (jiridanek) [inline base/ubi9-python-3.11/Dockerfile:1]: undo this, did that just to trigger pr tests

File: .github/workflows/build-notebooks-TEMPLATE.yaml [hunk 1]
```diff
@@ -46,6 +46,8 @@ jobs:
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
+      # region Free up disk space
```
Comment (jiridanek) [inline .github/workflows/build-notebooks-TEMPLATE.yaml:49]: This is the VS Code-compatible syntax for collapsable regions. Allows to collapse parts of the file in IntelliJ and VS Code

image

**Comment by atheo89** [comment]
Great work Jiri! 
/lgtm

**Comment by jiridanek** [comment]
/override [ci/prow/images](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/opendatahub-io_notebooks/775/pull-ci-opendatahub-io-notebooks-main-images/1862095289750392832)

**Comment by jiridanek** [comment]
/override build (rocm-jupyter-pytorch-ubi9-python-3.9) / build

**Comment by jiridanek** [comment]
/override "build (rocm-jupyter-pytorch-ubi9-python-3.9) / build"

**Comment by jiridanek** [comment]
/override "build (rocm-jupyter-pytorch-ubi9-python-3.11) / build"

**Comment by jiridanek** [comment]
/override ci/prow/notebook-rocm-jupyter-pyt-ubi9-python-3-11-pr-image-mirror

**Comment by jiridanek** [comment]
/override ci/prow/notebook-rocm-jupyter-pyt-ubi9-python-3-9-pr-image-mirror
/override ci/prow/notebook-rocm-jupyter-tf-ubi9-python-3-11-pr-image-mirror
/override ci/prow/notebook-rocm-jupyter-tf-ubi9-python-3-9-pr-image-mirror

File: jupyter/rocm/pytorch/ubi9-python-3.11/kustomize/base/kustomization.yaml [hunk 2]
```diff
@@ -1,9 +1,9 @@
 ---
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
-namePrefix: rocm-jupyter-pytorch-ubi9-python-3-11-
+namePrefix: jupyter-rocm-pytorch-ubi9-python-3-11-
 commonLabels:
-  app: rocm-jupyter-pytorch-ubi9-python-3-11
+  app: jupyter-rocm-pytorch-ubi9-python-3-11
```
Comment (jstourac) [inline jupyter/rocm/pytorch/ubi9-python-3.11/kustomize/base/kustomization.yaml:6]: Hey, didn't check - what does this affects? Is it safe? :thinking:

Comment (jiridanek) [reply jupyter/rocm/pytorch/ubi9-python-3.11/kustomize/base/kustomization.yaml:6]: does not run without this change in gha, and I did not manage to make the tests run in openshift-ci, so, yeah, I think it makes things better overall

**Pull request merged**